File: insert_delete_get_random_o(1).cpp
class RandomizedSet {
private:
    unordered_map<int, int> mp;
    vector<int> v;
    int size;
    
public:
    RandomizedSet() {
        size = 0;
    }
    
    bool insert(int val) {
        if(mp.find(val) == mp.end()) {
            mp[val] = size++;
            v.push_back(val);
            return true;
        }
        return false;
    }
    
    bool remove(int val) {
        if(mp.find(val) != mp.end()) {
            int last = v[--size];
            v.pop_back();

            v[mp[val]] = last;
            mp[last] = mp[val];
            
            mp.erase(val);
            return true;
        }
        return false;
    }
    
    int getRandom() {
        int random = rand();
        int randIndex = random % size;
        return v[randIndex];
    }
};

// Time Complexity : O(1)
// Space Complexity : O(n)

File: insert_delete_get_random_o(1).go
type RandomizedSet struct {
    mp map[int]int
    vec []int
    size int
}


func Constructor() RandomizedSet {
    mp := make(map[int]int)
    vec := make([]int, 0)
    return RandomizedSet{mp, vec, 0}
}


func (this *RandomizedSet) Insert(val int) bool {
    _, ok := this.mp[val];
    if ok {
        return false
    }
    this.mp[val] = this.size
    this.size = this.size+1
    this.vec = append(this.vec, val)
    return true
}


func (this *RandomizedSet) Remove(val int) bool {
    _, ok := this.mp[val]; 
    if !ok {
        return false
    }

    i := this.mp[val]
    this.size = this.size-1
    
    this.vec[i] = this.vec[this.size]
    this.vec[this.size] = this.vec[i]
    this.mp[this.vec[i]] = i
    this.vec = this.vec[:this.size]
    delete(this.mp, val)
    return true
}


func (this *RandomizedSet) GetRandom() int {
    i := rand.Intn(this.size)
    return this.vec[i]
}

// Time Complexity: O(1)
// Space Complexity: O(n)
